pkg/controllers/consolidation/stabilizationwindow.go

```diff
@@ -0,0 +1,201 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package consolidation
+
+import (
+	"context"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter/pkg/controllers/state"
+	"github.com/aws/karpenter/pkg/utils/pod"
+)
+
+type StabilizationWindow struct {
+	clock clock.Clock
+
+	lastConsolidationTime time.Time
+	cluster               *state.Cluster
+	kubeClient            client.Client
+}
+
+// fastWindowTime is the  window used if all the cluster workloads are reporting ready
+const fastWindowTime = 15 * time.Second
```

Comment:
I wonder if this is large enough that we'll see significant slow down for large scale downs. IIUC, all voluntary terminations will eventually be subject to this window, and this could really add up. 

IIUC, there are two cases here. 1/ allow controllers to recreate pods, and 2/, cover the delete -> create deployment case. I'm a bit hesitant of optimizing too heavily towards the latter case. WDYT about shortening this to something like 5 seconds?

Reply:
5 seconds is probably ok, trying to balance letting controllers react but since we trigger on pod deletion I think it will be fine.